### ISSUE TITLE
Pnpm not working with Expo

### ISSUE THREAD

**Issue by MariuzM** [expo/expo#22413]
### Minimal reproducible example

https://github.com/MariuzM/expo-test

### Summary

When i run `pnpm ios` i get his error. The Expo install is latest with typescript. No other tweaks / changes have been done. 

![image](https://user-images.githubusercontent.com/36937039/236819900-7a57f21a-2562-42f9-8cca-0d68e56134fa.png)

### Environment

Local

**Comment by LeDominik** [comment]
I'm also trying to get pnpm working -- what holds me back is that `expo-modules` (bootstrapped cleanly from the expo modules template) plain out refuses to use pnpm... otherwise I guess it would work following all the fixes in here: https://github.com/byCedric/expo-monorepo-example

From what I can observe around [metro#1](https://github.com/facebook/metro/issues/1) we're very close to this actually all working nicely with just plain (space-efficient & fast) pnpm symlinks.

**Comment by LeDominik** [comment]
@MariuzM  you'd at least need an `.npmrc` to get a flat `node_module` directory:

```
node-linker=hoisted
```

**Comment by karlhorky** [comment]
@MariuzM should this be reopened now that [Expo SDK 50](https://expo.dev/changelog/2024/01-18-sdk-50) is out, which uses [React Native 0.73](https://reactnative.dev/blog/2023/12/06/0.73-debugging-improvements-stable-symlinks) internally?

React 0.73 has made symlink support stable. Would be great to get `pnpm` symlink support in Expo apps (without `node-linker=hoisted` or any other workarounds).

**Comment by MariuzM** [comment]
Hey, hm not sure i've stoped using PNPM and moved to Bun

**Comment by karlhorky** [comment]
Just tried it with a new `create-expo` app just now (without using `.npmrc` with `node-linker=hoisted`, and it's still a problem:

```bash
$ mkdir a
$ cd a
$ pnpm create expo@latest --template expo-template-blank-typescript .

.../Library/pnpm/store/v3/tmp/dlx-68676  |   +1 +
.../Library/pnpm/store/v3/tmp/dlx-68676  | Progress: resolved 1, reused 0, downloaded 1, added 1, done
✔ Downloaded and extracted project files.
> pnpm install
 WARN  8 deprecated subdependencies found: @babel/plugin-proposal-async-generator-functions@7.20.7, @babel/plugin-proposal-class-properties@7.18.6, @babel/plugin-proposal-nullish-coalescing-operator@7.18.6, @babel/plugin-proposal-numeric-separator@7.18.6, @babel/plugin-proposal-object-rest-spread@7.20.7, @babel/plugin-proposal-optional-catch-binding@7.18.6, @babel/plugin-proposal-optional-chaining@7.21.0, @npmcli/move-file@1.1.2
Packages: +840
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 847, reused 833, downloaded 7, added 840, done

dependencies:
+ expo 50.0.4
+ expo-status-bar 1.11.1
+ react 18.2.0
+ react-native 0.73.2

devDependencies:
+ @babel/core 7.23.9
+ @types/react 18.2.48
+ typescript 5.3.3

Done in 17.7s

✅ Your project is ready!

To run your project, run one of the following pnpm commands.

- pnpm run android
- pnpm run ios
- pnpm run web
```

The app scaffolded just fine, but as soon as I tried to run it on an Android emulator, it crashed, unable to resolve `../../App`:

```
pnpm run android

> a@1.0.0 android /Users/k/p/a
> expo start --android

Starting project at /Users/k/p/a
Starting Metro Bundler
› Opening ... on Pixel_3a_API_30
✔ Expo Go on Pixel_3a_API_30 is outdated, would you like to upgrade? … yes
Uninstalling Expo Go from android device Pixel_3a_API_30.
Downloading the Expo Go app [================================================================] 100% 0.0s

...QR Code...

› Metro waiting on ...
› Scan the QR code above with Expo Go (Android) or the Camera app (iOS)

› Using Expo Go
› Press s │ switch to development build

› Press a │ open Android
› Press i │ open iOS simulator
› Press w │ open web

› Press j │ open debugger
› Press r │ reload app
› Press m │ toggle menu
› Press o │ open project code in your editor

› Press ? │ show all commands

Logs for your project will appear below. Press Ctrl+C to exit.
Android Bundling failed 710ms (node_modules/.pnpm/expo@50.0.4_@babel+core@7.23.9_@react-native+babel-preset@0.73.20/node_modules/expo/AppEntry.js)
Unable to resolve "../../App" from "node_modules/.pnpm/expo@50.0.4_@babel+core@7.23.9_@react-native+babel-preset@0.73.20/node_modules/expo/AppEntry.js"
```

It seems to be because of some hardcoded relative path in `expo/AppEntry.js` (which is deeper in the `node_modules` directory structure, because of how `pnpm` organizes modules.

**Comment by karlhorky** [comment]
Also asked Expo and pnpm maintainers over here in the open `pnpm` issue about whether it would be possible to get symlinks support out of the box:

- https://github.com/pnpm/pnpm/issues/4286#issuecomment-1913173223

**Comment by karlhorky** [comment]
Not stale

**Comment by devlopersabbir** [comment]
with `bun` it works perfectly 🥲

**Comment by YuriiLutsykSPRO** [comment]
pnpm works with expo

**Comment by karlhorky** [comment]
It appears that a community member @Zn4rK has figured out how to use Expo and Expo Router with pnpm symlinks (without `node-linker=hoisted` in `.npmrc`) 👀 🔥

- https://github.com/pnpm/pnpm/issues/4286#issuecomment-2218470271

**Comment by karlhorky** [comment]
Looks like non-hoisted pnpm support is scheduled for Expo SDK 52 - mentioned in @byCedric's post in the pnpm repo:

- https://github.com/pnpm/pnpm/issues/4286#issuecomment-2233017605

**Comment by karlhorky** [comment]
not stale

**Comment by CoRExE** [comment]
It's really sad
Cannot find module 'react'

<img width="296" alt="Image" src="https://github.com/user-attachments/assets/78b23377-dcee-477c-b8c9-c90595be1603" />

I have the .npmrc
and here my package.json :

[package.json](https://github.com/user-attachments/files/19272031/package.json)

Here the joke, you have command on Doc to create project and the project without any modification don't work

**Comment by CoRExE** [comment]
Same error with Bun

**Comment by CoRExE** [comment]
I change "output": "static" to "output": "single" in app.json file here and `pnpx expo-doctor` and i have a result

**Comment by karlhorky** [comment]
The Expo SDK 53 release notes also didn't include anything about non-hoisted pnpm support:

- https://expo.dev/changelog/sdk-53

**Comment by dsifriend** [comment]
Still a problem…

**Comment by rafinskipg** [comment]
fuck expo in monorepos, what the hell, we are in 2025 monorepos should work

**Comment by apexglory** [comment]
1.Create index.tsx in Root:

import { registerRootComponent } from 'expo';
import App from './App';

registerRootComponent(App);

2.Change package.json:

  "main": "./index.tsx",

**Comment by OzgurBurakOzdemir** [comment]
With Expo SDK 53, after setting node-linker=hoisted, simply delete node_modules and pnpm-lock.yaml, then run pnpm install.

**Comment by karlhorky** [comment]
Seems like Expo SDK 54 may include full pnpm support without `node-linker=hoisted` in `.npmrc` 👀

> > @ssijak: I can safely use expo app in a pnpm turborepo monorepo without hoisting issues?
>
> @kitten: yep, should work as intended. A lot was already working (see @byCedric's https://github.com/byCedric/expo-monorepo-example) but isolated-installation doesn't have to be turned off anymore

- https://x.com/_philpl/status/1960754610664087981

**Comment by cusxio** [comment]
@karlhorky 

I tested isolated installs with `pnpm` on my project using SDK 54, and everything worked perfectly.

However, it doesn’t work with `bun`, Expo seems unable to locate some babel-related files, though I’m not sure why.

**Comment by karlhorky** [comment]
Yes, according to @byCedric's monorepo example repo, [Expo SDK 54](https://expo.dev/changelog/sdk-54) has support for pnpm without the `.npmrc` workaround of `node-linker=hoisted` 🎉🔥🚀

- Upgrade to Expo SDK 54 https://github.com/byCedric/expo-monorepo-example/pull/154
- Switch to pnpm isolated modules https://github.com/byCedric/expo-monorepo-example/pull/152
- Removal of the pnpm `metro.config.js` changes https://github.com/byCedric/expo-monorepo-example/commit/7892cfc4df02c5c29b1c137cd2ff64d45b0306ae

So I guess this issue can be closed now? @MariuzM @byCedric do you agree?

**Comment by kitten** [comment]
Yep, this should be closed. Supported in Expo SDK 54.
